apps/aechannel/src/aesc_utils.erl

```diff
@@ -479,8 +479,8 @@ process_force_progress(Tx, Addresses,
     PoITrees0 = trees_from_poi(Addresses, PoI),
     Reserve = aesc_channels:channel_reserve(Channel),
     PoITrees =
-        try aesc_offchain_update:apply_on_trees(Update, PoITrees0, NextRound,
-                                                      Reserve)
+        try aesc_offchain_update:apply_on_trees(Update, PoITrees0, Trees, Env,
```

Comment:
`Fee` must be charged upfront. Please amend, or track ticket analogous to https://www.pivotaltracker.com/story/show/160669401 , or at least add comment in same mentioned ticket.

Reply:
Checking this better, this is two questions:
* (A) Is state hash generated by off-chain update call contract, and meant to be compared with specified state hash, meant to include charging of fee?
* (B) Is fee in force progress transaction meant to be checked as available in the balance (alongside gaslimit * gasprice) before processing force progress transactions?

B seems done correctly - in the `check` callback.

So the question is A.

Reply:
`Fee` is the amount being charged from the `from` (the one that posts the force progress on-chain) and awared to the miner for accepting the transaction. Amount is being charged from the `from`'s on-chain account. This does not impact off-chain trees in any way and thus - has no impact on their root hash.

Reply:
> Do I understand correctly that the off-chain update can use only the balance of the initiator and responder that is the off-chain trees i.e. balance that has already been taken out of the on-chain trees? If so, it seems correct that the fee does not impact the state hash of the off-chain update. I.e. correct as implemented already so no changes needed.

So this is a little bit tricky. At channel creation both participants can dedicate some tokens. From then on - those are locked inside the channel. This is the `channel_amount` and can be only modified by deposits and withdrawals.

Inside the channel this `channel_amount` can be devided between participants or accounts of their desire - those could be some further accounts that only live off-chain, for example off-chain contracts.

Channel off-chain contract calls can modify all off-chain balances. They can NOT interact with on-chain balances neither in off-chain calls, nor via a forced progress call.
So yes, off-chain calls can only update off-chain balances but those could be `responder`, `initator` or some other off-chain accounts as well.